README.md
candy-pi-lite-os-image
===

Raspbian OS image builder for [CANDY Pi Lite Board](https://translate.google.com/translate?sl=auto&tl=en&js=y&prev=_t&hl=en&ie=UTF-8&u=https%3A%2F%2Fwww.candy-line.io%2F製品一覧%2Fcandy-pi-lite%2F&edit-text=&act=url)

# Supported Raspberry Pi

ARMv6/ARMv7 boards are available for this image.

- [Raspberry Pi Zero (Pin Header is required)](https://www.raspberrypi.org/products/raspberry-pi-zero/)
- [Raspberry Pi Zero W (Pin Header is required)](https://www.raspberrypi.org/products/raspberry-pi-zero/)
- [Raspberry Pi Zero WH](https://www.raspberrypi.org/blog/zero-wh/)
- [Raspberry Pi 1 Model A+](https://www.raspberrypi.org/products/raspberry-pi-1-model/)
- [Raspberry Pi 1 Model B+](https://www.raspberrypi.org/products/raspberry-pi-1-model-b/)
- [Raspberry Pi 2 Model B](https://www.raspberrypi.org/products/raspberry-pi-2-model-b/)
- [Raspberry Pi 3 Model A+](https://www.raspberrypi.org/products/raspberry-pi-3-model-a-plus/)
- [Raspberry Pi 3 Model B](https://www.raspberrypi.org/products/raspberry-pi-3-model-b/)
- [Raspberry Pi 3 Model B+](https://www.raspberrypi.org/products/raspberry-pi-3-model-b-plus/)
- [Raspberry Pi 4 Model B](https://www.raspberrypi.org/products/raspberry-pi-4-model-b/)

# Raspbian OS Version

- Raspbian Buster Version September 2019 (Release date: 2019-09-26, Kernel: v4.19) with some library updates at the time of the image creation

# Features

- [Raspbian OS Lite (Raspbian Buster Lite)](https://www.raspberrypi.org/downloads/raspbian/) based image with software updates
- [CANDY Pi Lite Board Service](https://github.com/CANDY-LINE/candy-pi-lite-service) is installed
- [CANDY RED](https://github.com/CANDY-LINE/candy-red), a Node-RED based software dedicated to [CANDY Pi Lite Board](https://translate.google.com/translate?sl=auto&tl=en&js=y&prev=_t&hl=en&ie=UTF-8&u=https%3A%2F%2Fwww.candy-line.io%2F製品一覧%2Fcandy-pi-lite%2F&edit-text=&act=url) and [CANDY EGG Cloud Service](https://translate.google.com/translate?hl=en&sl=ja&tl=en&u=https%3A%2F%2Fwww.candy-line.io%2F製品一覧%2Fcandy-red-egg%2F), is installed
- `ufw` is enabled (denying from all traffic on `ppp0` and `wwan0`)
- Hardware watchdog is enabled
- Node.js v12 (armv6l) is installed (in Active LTS)
- `ssh` is **DISABLED** by default. Place an empty file as `/boot/ssh` to enable it
- Extra Debian Packages
  - `git`
  - `screen`
  - `fswebcam`
  - `sense-hat`
- Extra Python Packages
  - `pillow`
  - `pyserial`

## Distributions

- `*-en_US` suffix image contains US Keyboard Layout with UTC timezone and en_US.UTF-8 locale
- `*-ja_JP` suffix image contains Japanese Keyboard Layout with JST timezone and ja_JP.UTF-8 locale and CJK fonts

# How to flash the image

Use [balenaEtcher](https://www.balena.io/etcher/) for burning the image, which is a cross-platform app.

# How to build images

```
# create a tag for the version then push it to remote

# using IIJ Mobile
FIRST_USER_NAME="pi" \
  FIRST_USER_PASS="raspberry" \
  ENABLE_SSH="0" \
  BOOT_APN="iijmobile.biz-ipv4v6" \
  ./scripts/before_script.sh

# using soracom.io
FIRST_USER_NAME="pi" \
  FIRST_USER_PASS="raspberry" \
  ENABLE_SSH="0" \
  BOOT_APN="soracom.io" \
  ./scripts/before_script.sh

# Let's build an image
time ./scripts/build_img.sh

# Upload image files
RELEASE_VERSION=16.0.0

GITHUB_OAUTH_TOKEN=YOUR_GITHUB_OAUTH_TOKEN \
  TAG_NAME=${RELEASE_VERSION} \
  ./scripts/deploy_zip.sh
```

# Trouble Shooting

```
Previous stage rootfs not found
******** Failed to build image files... ********
```
Remove empty directories under `pi-gen` directory.

# How to release

1. Edit `IMG_VERSION`, `CANDY_PI_LITE_VERSION`, `CANDY_RED_HASH`, `RPI_FIRMWARE_VERSION` and `NODEJS_VERSION` in before_script.sh
1. Create image files
1. Test if they're working
1. Upload image files and package info
1. Tag the version


CHANGELOG.md
# Revision History
* 16.0.0
  - CANDY Pi Lite Board Service 10.4.0
  - CANDY RED 9.11.0
  - Update Base Raspbian Buster version to `2021-05-07-raspbian-buster`
  - Raspberry Pi OS Firmware v[1.20210831](https://github.com/raspberrypi/firmware/tree/1.20210831)
    - Linux Kernel 5.10

* 15.0.0
  - CANDY Pi Lite Board Service 10.3.0
  - CANDY RED 9.11.0
  - Raspberry Pi OS Firmware v[1.20210527](https://github.com/raspberrypi/firmware/tree/1.20210527)
    - Linux Kernel 5.10

* 14.0.0
  - Node.js v12.22.1
  - CANDY RED 9.9.3
  - Update Base Raspbian Buster version to `2021-03-04-raspbian-buster`
  - Raspberry Pi OS Firmware v[1.20210303](https://github.com/raspberrypi/firmware/tree/1.20210303)
    - Linux Kernel 5.10

* 13.1.0
  - Node.js v12.20.1
  - CANDY RED 9.9.3
  - Add a file containing the published OS image version

* 13.0.0
  - CANDY Pi Lite Board Service 10.2.1
  - CANDY RED 9.9.2
  - Update Base Raspbian Buster version to `2020-12-02-raspbian-buster`
  - Raspberry Pi OS Firmware v[1.20210108](https://github.com/raspberrypi/firmware/tree/1.20210108)
    - Linux Kernel 5.4

* 12.0.0
  - CANDY Pi Lite Board Service 10.2.1
  - CANDY RED 9.9.0
  - Raspberry Pi OS Firmware v[1.20201022](https://github.com/raspberrypi/firmware/tree/1.20201022)
    - Linux Kernel 5.4

* 11.1.0
  - CANDY Pi Lite Board Service 10.1.0
  - CANDY RED 9.8.1

* 11.0.0
  - CANDY Pi Lite Board Service 10.0.1
  - CANDY RED 9.7.0
  - Raspberry Pi OS Firmware v[1.20200819](https://github.com/raspberrypi/firmware/tree/1.20200819)
    - Linux Kernel 5.4

* 10.0.0
  - CANDY Pi Lite Board Service 10.0.1
  - CANDY RED 9.6.0
  - Raspberry Pi OS Firmware v[1.20200723](https://github.com/raspberrypi/firmware/tree/1.20200723)
    - Linux Kernel 5.4

* 9.4.1
  - Fix the image script

* 9.4.0
  - CANDY Pi Lite Board Service 9.0.0
  - CANDY RED 9.6.0
  - Bump Node.js version to v12.18.3
  - Raspberry Pi OS Firmware v[1.20200601](https://github.com/raspberrypi/firmware/tree/1.20200601)

* 9.3.2
  - CANDY Pi Lite Board Service 8.0.3
  - CANDY RED 9.5.0
  - Add a new config for host name
  - Fix release note link

* 9.3.1
  - CANDY Pi Lite Board Service 8.0.1

* 9.3.0
  - Bump Node.js version to v12.16.3
  - Remove pip installation for pillow pyserial pckages
  - BOOT_APN is now manfatory
  - CANDY Pi Lite Board Service 8.0.0
  - CANDY RED 9.4.0

* 9.2.0
  - Add `fbterm` package for ja-JP image
  - Update Base Raspbian Buster version to `2020-02-13-raspbian-buster`
  - Bump Node.js version to v12.16.1

* 9.1.0
  - CANDY Pi Lite Board Service 7.1.1
  - CANDY RED 9.2.0

* 9.0.0
  - CANDY Pi Lite Board Service 7.1.0
  - CANDY RED 9.1.0
  - ~~Update Base Raspbian Buster version to `2020-02-13-raspbian-buster`~~
  - Bump Node.js version to v12.16.1

* 8.0.0
  - Update Base Raspbian Buster version to `2019-09-26-raspbian-buster`

* 7.2.0
  - CANDY Pi Lite Board Service 7.0.3
  - CANDY RED 9.0.2
  - Bump Node.js version to v12.14.1

* 7.1.0
  - CANDY Pi Lite Board Service 7.0.2
  - CANDY RED 9.0.0
  - Bump Node.js version to v12.14.0

* 7.0.1
  - CANDY Pi Lite Board Service 7.0.2
  - CANDY RED 8.3.2
  - Bump Node.js version to v10.18.0
  - Fix a qemu issue

* 7.0.0
  - Update Base Raspbian Buster version to `2019-07-10-raspbian-buster`
    - **Debian Buster is introduced as of this release!**
  - CANDY Pi Lite Board Service 7.0.1
  - CANDY RED 8.3.0
  - Bump Node.js version to v10.15.3
  - CJK fonts are installed for ja_JP image

* 6.0.2
  - CANDY Pi Lite Board Service 6.6.0
  - CANDY RED 8.2.2

* 6.0.1
  - CANDY Pi Lite Board Service 6.5.0
  - CANDY RED 8.2.2

* 6.0.0
  - Update Base Raspbian Stretch version to `2018-11-13-raspbian-stretch`
  - CANDY Pi Lite Board Service 6.4.0
  - CANDY RED 8.2.0
  - Bump Node.js version to v8.15.0

* 5.1.0
  - CANDY Pi Lite Board Service 6.2.0
  - CANDY RED 8.1.0

* 5.1.0
  - CANDY Pi Lite Board Service 6.2.0
  - CANDY RED 8.0.2

* 5.0.0
  - Update Base Raspbian Stretch version to `2018-10-09-raspbian-stretch`
  - CANDY Pi Lite Board Service 6.1.0
  - CANDY RED 8.0.1
  - Bump Node.js version to v8.12.0

* 4.2.1
  - CANDY RED 8.0.1 with node-red-contrib-generic-ble update (v3.0.0)

* 4.2.0
  - CANDY Pi Lite Board Service 6.0.0

* 4.1.0
  - CANDY Pi Lite Board Service 5.2.0
  - CANDY RED 8.0.1
  - Bump Node.js version to v8.11.4

* 4.0.0
  - Update Base Raspbian Stretch version to `2018-06-27-raspbian-stretch`
  - CANDY Pi Lite Board Service 5.1.0
  - CANDY RED 7.1.0
  - Bump Node.js version to v8.11.3

* 3.1.1
  - CANDY Pi Lite Board Service 5.0.1
  - CANDY RED 7.0.2

* 3.1.0
  - CANDY Pi Lite Board Service 5.0.0
  - CANDY RED 7.0.0
  - Add PiCamera software
  - Bump Node.js version to v6.14.2

* 3.0.0
  - Update Base Raspbian Stretch version to `2018-03-13-raspbian-stretch`
  - CANDY Pi Lite Board Service 4.0.0
  - CANDY RED 6.1.0
  - Add `hostapd` and `dnsmasq` (disabled by default)
  - Bump Node.js version to v6.14.1

* 2.1.4
  - Fix an issue where `ufw` was inactive

* 2.1.3
  - CANDY Pi Lite Board Service 2.0.2

* 2.1.2
  - Remove nodejs package

* 2.1.1
  - CANDY Pi Lite Board Service 2.0.1

* 2.1.0
  - ARMv6 Raspberry Pi Boards(Zero/Zero W/Zero WH/Model A+/Model B+) Support
  - CANDY Pi Lite Board Service 2.0.0
  - CANDY RED 5.6.1

* 2.0.1
  - CANDY Pi Lite Board Service 1.8.2
  - CANDY RED 5.6.1

* 2.0.0
  - Update Base Raspbian Stretch version to `2017-11-29-raspbian-stretch`
  - CANDY Pi Lite Board Service 1.8.0
  - CANDY RED 5.6.1

* 1.3.0
  - CANDY Pi Lite Board Service 1.7.3
  - CANDY RED 5.5.0

* 1.2.0
  - CANDY Pi Lite Board Service 1.7.2
  - CANDY RED 5.5.0

* 1.1.0
  - Sense HAT support
  - Improve package installation time

* 1.0.0
  - Initial Release


scripts/before_script.sh
#!/usr/bin/env bash

IMG_VERSION="v16.0.0"
DEPS_DIR=$(pwd)/deps
PIGEN_DIR=${DEPS_DIR}/pi-gen
MOVIDIUS=${MOVIDIUS:-0}
CANDY_PI_LITE_VERSION=${CANDY_PI_LITE_VERSION:-"10.4.0"}
CANDY_RED_HASH=${CANDY_RED_HASH:-"9.11.0"}
RPI_FIRMWARE_VERSION=${RPI_FIRMWARE_VERSION:-"1.20210831"}
NODEJS_VERSION="12.22.6"
BOOT_APN=${BOOT_APN:-""}
BUTTON_EXT=${BUTTON_EXT:-""}
BASE_ROOT_MARGIN=${BASE_ROOT_MARGIN:-"400"}

function clean_setup {
  if [ "${NO_CLEAN}" != "1" ]; then
    git submodule update --init --recursive
    pushd ${PIGEN_DIR}
    git reset --hard HEAD
    git clean -f -d
    popd
  fi
}

function add_config {
  if [ -n "${!1}" ]; then
    echo "[add_config] ${1} => [${!1}]"
    echo "export $1=${!1}" >> ${PIGEN_DIR}/config
  fi
}

function create_config {
  IMG_NAME="${IMG_NAME:-candy-pi-lite-raspbian}"
  echo "IMG_NAME=${IMG_NAME}" > ${PIGEN_DIR}/config
  pushd ${PIGEN_DIR}
  GIT_HASH=$(git rev-parse HEAD)
  popd
  add_config GIT_HASH
  add_config FIRST_USER_NAME
  add_config FIRST_USER_PASS
  add_config ENABLE_SSH
  add_config TARGET_HOSTNAME
  add_config IMG_VERSION
  add_config NODEJS_VERSION
}

function configure_stages {
  touch ${PIGEN_DIR}/stage{3,4,5}/SKIP
  rm -f ${PIGEN_DIR}/stage{2,3,4,5}/EXPORT_NOOBS
  rm -f ${PIGEN_DIR}/stage{2,3,4,5}/EXPORT_IMAGE
  pushd $(pwd)/overlay
  for STAGE in $(ls -d stage*); do
    cp -fr ${STAGE} ../deps/pi-gen/
  done
  popd
  sed -i -e "s/raspberrypi-kernel/device-tree-compiler/g" ${PIGEN_DIR}/stage0/02-firmware/01-packages
  sed -i -e "s/RPI_FIRMWARE_VERSION=/RPI_FIRMWARE_VERSION=${RPI_FIRMWARE_VERSION}/g" ${PIGEN_DIR}/stage0/02-firmware/01-run.sh
  sed -i -e "s/libraspberrypi-bin libraspberrypi0 raspi-config/raspi-config/g" ${PIGEN_DIR}/stage1/03-install-packages/00-packages
  sed -i -e "s/libraspberrypi-dev libraspberrypi-doc libfreetype6-dev/libfreetype6-dev/g" ${PIGEN_DIR}/stage2/01-sys-tweaks/00-packages
  sed -i -e "s/RPI_FIRMWARE_VERSION=/RPI_FIRMWARE_VERSION=${RPI_FIRMWARE_VERSION}/g" ${PIGEN_DIR}/stage0/02-firmware/01-run-chroot.sh
  sed -i -e "s/raspberrypi-kernel/device-tree-compiler/g" ${PIGEN_DIR}/stage0/02-firmware/01-packages
  sed -i -e "s/ 0\.2 \+ 200 \* 1024 \* 1024/ 0.2 + ${BASE_ROOT_MARGIN} * 1024 * 1024/g" ${PIGEN_DIR}/export-image/prerun.sh
}

function configure_scripts {
  sed -i -e "s/CANDY_PI_LITE_VERSION=/CANDY_PI_LITE_VERSION=${CANDY_PI_LITE_VERSION}/g" ${PIGEN_DIR}/stage2-1-en_US/99-candy-pi-lite/00-run-chroot.sh
  sed -i -e "s/CANDY_RED_HASH=/CANDY_RED_HASH=${CANDY_RED_HASH}/g" ${PIGEN_DIR}/stage2-1-en_US/99-candy-pi-lite/00-run-chroot.sh
  if [ -n "${BOOT_APN}" ]; then
    sed -i -e "s/BOOT_APN=iijmobile.biz-ipv4v6/BOOT_APN=${BOOT_APN}/g" ${PIGEN_DIR}/stage2-1-en_US/99-candy-pi-lite/00-run-chroot.sh
    echo "[INFO] BOOT_APN => ${BOOT_APN}"
  else
    echo "[ERROR] BOOT_APN is now mandatory."
    exit 1
  fi
  if [ -n "${BUTTON_EXT}" ]; then
    sed -i -e "s/BUTTON_EXT=0/BUTTON_EXT=${BUTTON_EXT}/g" ${PIGEN_DIR}/stage2-1-en_US/99-candy-pi-lite/00-run-chroot.sh
  fi
}

function apply_macos_support {
  if [ `uname` = "Darwin" ]; then
    # Use sed -E rather than sed -r
    sed -i -e "s/sed -r/sed -E/g" ${PIGEN_DIR}/build-docker.sh
    # Use a python code rather than realpath
    PYTHON="python"
    PYTHON_TEST=`${PYTHON} -V`
    if [ "$?" != "0" ]; then
      PYTHON="python3"
      PYTHON_TEST=`${PYTHON} -V`
      if [ "$?" != "0" ]; then
        echo "[ERROR] Python is missing"
        exit 1
      fi
    fi
    sed -i -e "s/CONFIG_FILE=\$(realpath -s \"\$CONFIG_FILE\")/CONFIG_FILE=\$(${PYTHON} -c \"import os\; print(os.path.realpath('\$CONFIG_FILE'))\")/g" ${PIGEN_DIR}/build-docker.sh
  fi
}

function use_32bit_image {
  # in order to apply the workaround for a qemu issue
  # https://github.com/RPi-Distro/pi-gen/issues/271#issuecomment-556812205
  sed -i -e "s/FROM debian:buster/FROM i386\\/debian:buster/g" ${PIGEN_DIR}/Dockerfile
}

clean_setup
create_config
configure_stages
configure_scripts
use_32bit_image
apply_macos_support
